File: game-server/src/main/java/com/kingston/mmorpg/game/buff/model/Buff.java
package com.kingston.mmorpg.game.buff.model;

import java.util.concurrent.Future;

import com.kingston.mmorpg.game.base.SpringContext;
import com.kingston.mmorpg.game.scene.actor.Creature;

public abstract class Buff {

	protected int modelId;

	/** 激活时间 */
	private long activeTime;

	/** 持续时长 */
	private long duration;

	private int overlapLayer;

	private Creature owner;

	private Creature origin;

	private Future<?> timeOutTask;

	public void init() {
		activeTime = System.currentTimeMillis();
		if (!isPermanent()) {
			registerTimeOutTask();
		}
		onStart();
	}

	public void onStart() {

	}

	public void registerTimeOutTask() {
		cancleTimeOutTask();
		timeOutTask = SpringContext.getSchedulerManager().schedule(new Runnable() {
			@Override
			public void run() {
				destroy();
			}
		}, getDuration());
	}

	public void cancleTimeOutTask() {
		if (timeOutTask != null) {
			timeOutTask.cancel(false);
			timeOutTask = null;
		}
	}

	/**
	 * 是否永久性buff
	 * 
	 * @return
	 */
	public boolean isPermanent() {
		return this.duration < 0;
	}

	public void destroy() {
		onDestroy();
	}

	public void onDestroy() {
		SpringContext.getBuffService().removeBuff(owner, this);
	}

	public int getModelId() {
		return modelId;
	}

	public void setModelId(int modelId) {
		this.modelId = modelId;
	}

	public long getActiveTime() {
		return activeTime;
	}

	public void setActiveTime(long activeTime) {
		this.activeTime = activeTime;
	}

	public long getDuration() {
		return duration;
	}

	public void setDuration(long duration) {
		this.duration = duration;
	}

	public int getOverlapLayer() {
		return overlapLayer;
	}

	public void setOverlapLayer(int overlapLayer) {
		this.overlapLayer = overlapLayer;
	}

	public Creature getOwner() {
		return owner;
	}

	public void setOwner(Creature owner) {
		this.owner = owner;
	}

	public Creature getOrigin() {
		return origin;
	}

	public void setOrigin(Creature origin) {
		this.origin = origin;
	}

}

File: game-server/src/main/kotlin/org/forfun/mmorpg/game/buff/model/Buff.kt
package org.forfun.mmorpg.game.buff.model

import io.micrometer.core.instrument.config.validate.PropertyValidator.getDuration
import org.forfun.mmmorpg.game.scene.actor.Creature
import org.forfun.mmorpg.game.base.GameContext
import java.util.concurrent.Future

abstract class Buff {

    var modelId: Int = 0

    /** 激活时间 */
    private var activeTime: Long = 0

    /** 持续时长 */
    var duration: Long = 0

    var overlapLayer: Int = 0

    var owner: Creature? = null

    var origin: Creature? = null

    private var timeOutTask: Future<*>? = null

    open fun init() {
        activeTime = System.currentTimeMillis()
        if (!isPermanent) {
            registerTimeOutTask()
        }
        onStart()
    }

    open fun onStart() {}

    fun registerTimeOutTask() {
        cancleTimeOutTask()
        timeOutTask = GameContext.schedulerManager().schedule({
            destroy()
        }, duration)
    }

    fun cancleTimeOutTask() {
        if (timeOutTask != null) {
            timeOutTask!!.cancel(false)
            timeOutTask = null
        }
    }

    /**
     * 是否永久性buff
     * @return
     */
    val isPermanent: Boolean
        get() = this.duration < 0

    fun destroy() {
        onDestroy()
    }

    open fun onDestroy() {
        GameContext.buffService().removeBuff(owner!!, this)
    }

    fun getActiveTime(): Long = activeTime

    fun setActiveTime(activeTime: Long) {
        this.activeTime = activeTime
    }

}
